File: app/src/main/java/com/stathis/elmepaunivapp/ui/department/recyclerviews/DepartmentAdapter.java
package com.stathis.elmepaunivapp.ui.department.recyclerviews;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.stathis.elmepaunivapp.R;
import com.stathis.elmepaunivapp.abstraction.DiffItemCallbackClass;
import com.stathis.elmepaunivapp.listeners.DepMembersClickListener;
import com.stathis.elmepaunivapp.listeners.DepartmentCardClickListener;
import com.stathis.elmepaunivapp.listeners.FieldsOfStudyListener;
import com.stathis.elmepaunivapp.listeners.ProgrammesClickListener;
import com.stathis.elmepaunivapp.listeners.SocialClickListener;
import com.stathis.elmepaunivapp.ui.department.model.EmptyModel;
import com.stathis.elmepaunivapp.ui.department.model.Research;
import com.stathis.elmepaunivapp.ui.department.model.VirtualTour;

public class DepartmentAdapter extends ListAdapter<Object, RecyclerView.ViewHolder> {

    FieldsOfStudyListener fieldsOfStudyListener;
    SocialClickListener socialClickListener;
    DepMembersClickListener depMembersClickListener;
    ProgrammesClickListener programmesClickListener;
    DepartmentCardClickListener departmentCardClickListener;

    public DepartmentAdapter(FieldsOfStudyListener fieldsOfStudyListener,SocialClickListener socialClickListener,
                             DepMembersClickListener depMembersClickListener,
                             ProgrammesClickListener programmesClickListener,DepartmentCardClickListener departmentCardClickListener) {
        super(new DiffItemCallbackClass<Object>());
        this.fieldsOfStudyListener = fieldsOfStudyListener;
        this.socialClickListener = socialClickListener;
        this.depMembersClickListener = depMembersClickListener;
        this.programmesClickListener = programmesClickListener;
        this.departmentCardClickListener = departmentCardClickListener;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if(viewType == R.layout.holder_empty_view){
            return new EmptyViewHolder(LayoutInflater.from(parent.getContext())
                    .inflate(viewType, parent, false));
        } else {
            return new DepartmentViewHolder(LayoutInflater.from(parent.getContext())
                    .inflate(viewType, parent, false));
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof DepartmentViewHolder) {
            ((DepartmentViewHolder) holder).present(getItem(position),fieldsOfStudyListener,socialClickListener,depMembersClickListener,programmesClickListener,departmentCardClickListener);
        } else {
            ((EmptyViewHolder) holder).present(getItem(position));
        }
    }

    @Override
    public int getItemViewType(int position) {
        if(getItem(position) instanceof Research){
            return R.layout.holder_department_research_item;
        } else if (getItem(position) instanceof VirtualTour){
            return R.layout.holder_virtual_tour;
        } else if (getItem(position) instanceof EmptyModel){
            return R.layout.holder_empty_view;
        } else {
            return R.layout.holder_research_item_row;
        }
    }
}

File: app/src/main/java/com/stathis/elmepaunivapp/ui/main/department/recyclerviews/DepartmentAdapter.kt
package com.stathis.elmepaunivapp.ui.main.department.recyclerviews

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.ListAdapter
import com.stathis.elmepaunivapp.R
import com.stathis.elmepaunivapp.abstraction.DiffItemClass
import com.stathis.elmepaunivapp.callbacks.ElmepaClickListener
import com.stathis.elmepaunivapp.abstraction.LocalModel
import com.stathis.elmepaunivapp.ui.main.department.model.EmptyModel
import com.stathis.elmepaunivapp.ui.main.department.model.HorizontalDepartmentItem
import com.stathis.elmepaunivapp.ui.main.department.model.VerticalDepartmentItem
import com.stathis.elmepaunivapp.ui.main.students.model.CarouselParent

class DepartmentAdapter(private val callback : ElmepaClickListener) : ListAdapter<LocalModel, DepartmentViewHolder>(DiffItemClass<LocalModel>()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DepartmentViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(viewType,parent,false)
        return DepartmentViewHolder(view, callback)

        //FIXME: Migrate to databinding pending
    }

    override fun onBindViewHolder(holder: DepartmentViewHolder, position: Int) {
        holder.bindData(getItem(position))
    }

    override fun getItemViewType(position: Int): Int = when(getItem(position)){
        is CarouselParent -> R.layout.holder_viewpager_item
        is HorizontalDepartmentItem -> R.layout.holder_parent_horizontal_nested_item
        is VerticalDepartmentItem -> R.layout.holder_parent_vertical_nested_item
        is EmptyModel -> R.layout.holder_empty_view
        else -> R.layout.holder_empty_view
    }
}